packages/flutter/lib/src/widgets/scroll_aware_image_provider.dart

```diff
@@ -0,0 +1,173 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/painting.dart';
+import 'package:flutter/scheduler.dart';
+
+import 'framework.dart';
+import 'scrollable.dart';
+
+/// Provides access to a [BuildContext] for a [ScrollAwareImageProvider].
+///
+/// Creators of this object must guarantee the following:
+///
+///   1. They create this object at or after [State.initState] but before
+///      [State.dispose]. In particular, do not attempt to create this from the
+///      constructor of a state.
+///   2. They call [dispose] from [State.dispose].
+///
+/// This object will not hold on to the [State] after disposal.
+@optionalTypeArgs
+class ScrollAwareContextProvider<T extends State> {
+  /// Creates a provider of a [BuildContext] for a [ScrollAwareImageProvider].
+  ///
+  /// Callers must call [dispose] when the [State] is disposed.
+  ///
+  /// The [State] must not be null, and [State.mounted] must be true.
+  ScrollAwareContextProvider(this._state)
+      : assert(_state != null),
+        assert(_state.mounted, 'A ScrollAwareContextProvider was given a BuildContext for an Element that was never mounted.');
+
+  /// Provides safe access to the build context.
+  ///
+  /// If [dispose] has been called, will return null.
+  ///
+  /// Otherwise, asserts the [_state] is still mounted and returns its context.
+  BuildContext get _context {
+    assert(debugValidate());
+    if (_state == null) {
+      return null;
+    }
+    return _state.context;
+  }
+
+  /// Called from asserts or tests to determine whether this object is in a
+  /// valid state.
+  ///
+  /// Always returns true, but will assert if [dispose] has not been called
+  /// but the state this is tracking is unmounted.
+  @visibleForTesting
+  bool debugValidate() {
+    assert(
+      _state == null || _state.mounted,
+      'A ScrollAwareContextProvider tried to access the BuildContext of a '
+      'disposed State object. This can happen when the creator of this provider '
+      'fails to call dispose when it is disposed.',
+    );
+    return true;
+  }
+
+  /// Do not touch this.  Use [_context].
+  T _state;
+
+  /// Marks the [BuildContext] as disposed.
+  ///
+  /// Creators of this object should call [dispose] when their [Element] is
+  /// unmounted, e.g. when [State.dispose] is called.
+  void dispose() {
+    _state = null;
+  }
+}
+
+/// An [ImageProvider] that makes use of
+/// [Scollable.recommendDeferredLoadingForContext] to avoid loading images when
+/// rapidly scrolling.
+///
+/// This provider assumes that its wrapped [imageProvider] correctly uses the
+/// [ImageCache], and does not attempt to re-acquire or decode images in the
+/// cache.
+///
+/// Calling [resolve] on this provider will cause it to obtain the image key
```

Comment:
Document somewhere what it means in practical terms for the image when the enumeration below says that resolve just "ends".